packages/w3p/src/providers/wrapped/evm-wallet-connect.ts

```diff
@@ -0,0 +1,316 @@
+      this.#defaultEventPayload,
+    )
+
+    this.emit(
+      this.isConnected
+        ? PROVIDER_EVENT_BUS_EVENTS.Connect
+        : PROVIDER_EVENT_BUS_EVENTS.Disconnect,
+      this.#defaultEventPayload,
+    )
+
+    this.#w3Modal.closeModal()
+  }
+
+  async disconnect() {
+    await this.#rawProvider.disconnect()
+    this.#chainId = undefined
+    this.#address = undefined
+    this.emit(PROVIDER_EVENT_BUS_EVENTS.Disconnect, this.#defaultEventPayload)
+  }
+
+  async #checkForPersistedSession() {
+    if (!this.#rawProvider) {
+      throw new ReferenceError('EthereumProvider is not initialized.')
+    }
+
+    if (!this.#rawProvider?.session) return
+
+    this.#chainId =
+      this.#rawProvider?.session?.namespaces?.eip155?.chains?.[0]?.split(':')[1]
+
+    this.#address =
+      this.#rawProvider?.session?.namespaces?.eip155?.accounts?.[0]?.split(
+        ':',
+      )?.[2]
+  }
+
+  getAddressUrl(chain: Chain, address: string): string {
+    return getEthExplorerAddressUrl(chain, address)
+  }
+
+  getTxUrl(chain: Chain, txHash: string): string {
+    return getEthExplorerTxUrl(chain, txHash)
+  }
+
+  getHashFromTx(txResponse: TransactionResponse): string {
+    return (txResponse as EthTransactionResponse).transactionHash
+  }
+
+  /**
+   * @description Switch the chain with WalletConnect Provider.
+   * You should always add a network first and then change it.
+   */
+
+  async switchChain(chainId: ChainId): Promise<void> {
+    if (!Provider.chainsDetails) {
+      throw new ReferenceError('Chains details are empty.')
+    }
+
+    const foundChain = Provider.chainsDetails[chainId]
+
+    if (!foundChain) {
+      throw new ReferenceError('Network is not supported')
+    }
+
+    await this.addChain(foundChain)
+
+    await requestSwitchEthChain(this.#ethProvider, foundChain.id)
+  }
+
+  async addChain(chain: Chain): Promise<void> {
+    await requestAddEthChain(this.#ethProvider, chain)
+  }
+
+  async signAndSendTx(tx: TxRequestBody): Promise<TransactionResponse> {
+    if (!this.#rawProvider) throw new TypeError(`Provider is not initialized`)
+
+    this.emit(PROVIDER_EVENT_BUS_EVENTS.BeforeTxSent, {
+      txBody: tx,
+    })
+
+    const transactionResponse = await this.#ethProvider?.send(
+      'eth_sendTransaction',
+      [tx],
+    )
+
+    this.emit(PROVIDER_EVENT_BUS_EVENTS.TxSent, {
+      txHash: transactionResponse?.hash,
+    })
+
+    const receipt = await transactionResponse?.wait()
+
+    this.emit(PROVIDER_EVENT_BUS_EVENTS.TxConfirmed, {
+      txResponse: receipt,
+    })
+
+    return receipt
+  }
+
+  async signMessage(message: string): Promise<string> {
+    if (!this.#rawProvider || !this.#address)
+      throw new TypeError(`Provider is not initialized`)
+
+    return this.#ethProvider?.send('personal_sign', [
+      utils.hexlify(utils.toUtf8Bytes(message)),
+      this.#address.toLowerCase(),
+    ])
+  }
+
+  async #setListeners() {
+    ;(this.#rawProvider as unknown as EthereumProvider).on(
```

Comment:
why a u casting type there if `this.#rawProvider.on` method works without it under this expression?

Reply:
done